fix(content-blog): links in feed should be absolute

VinceCYLiao opened the pull request:
… absolute

<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.
You can learn more about contributing to Docusaurus here: https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md
Happy contributing!
-->

## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

<!--
Please also remember to sign the CLA, although you can also sign it after submitting the PR. The CLA is required for us to merge your PR.
If this PR adds or changes functionality, please take some time to update the docs. You can also write docs after the API design is finalized and the code changes have been approved.
-->

## Motivation

Fix #9136

## Test Plan

added two mdx files in dogfood docs
website/_dogfooding/_blog tests/2023-07-19-a.mdx
website/_dogfooding/_blog tests/2023-07-19-b.mdx

Inside 2023-07-19-a.mdx are three links
```md
[absolute full url](https://github.com/facebook/docusaurus)

[absolute url with implicit domain name](/tests/blog/2023/07/19/b)

[relative url](2023-07-19-b.mdx)
```
Visit [/tests/blog/feed.json](https://deploy-preview-9151--docusaurus-2.netlify.app/tests/blog/feed.json)
1st link stays untouched 
2nd link resolved as "https://docusaurus.io/tests/blog/2023/07/19/b"
3rd link also resolved as "https://docusaurus.io/tests/blog/2023/07/19/b"
Which are correct.
![image](https://github.com/facebook/docusaurus/assets/46563912/bec66220-9539-4c27-95bb-3c164acab530)


### Test links

<!--
🙏 Please add an exhaustive list of links relevant to this pull request.
⏱ This saves maintainers a lot of time during reviews.

- If you changed anything that's displayed on UI, please add a dogfooding page in website/_dogfooding to help us preview the effect. Those tests are deployed at https://docusaurus.io/tests
- If you changed documentation, please link to the new and updated documentation pages.

After submission, our Netlify bot will post a deploy preview link in comment, in the format of https://deploy-preview-<PR-NUMBER>--docusaurus-2.netlify.app/. Once available, please edit this section with links to the relevant deploy preview pages.

Please don't be afraid to change the main site's configuration as well! You can make use of your new feature on our site so we can preview its effects. We can decide if it should be kept in production before merging it.
-->

Deploy preview:  https://deploy-preview-9151--docusaurus-2.netlify.app/blog/feed.json

## Related issues/PRs

[issue 9136](https://github.com/facebook/docusaurus/issues/9136)

Josh-Cena reviewed (changes requested):
The current solution is too error-prone because it includes too much custom logic. Plus it does not work with all relative links. Consider the following:

```html
<a href="another-post">link</a>
```

If the page is at `/blog/2020/09/13/current-post`, then the href should point to `/blog/2020/09/13/another-post`, not `/another-post`. I.e. the resolver needs to be aware of the current URL, not just the site's URL.

Also, I would prefer not manually joining URLs in any case. Why not `elm.attribs.href = String(new URL(elm.attribs.href, currentPageURL))`?

VinceCYLiao commented:
> The current solution is too error-prone because it includes too much custom logic. Plus it does not work with all relative links. Consider the following:
> 
> ```
> <a href="another-post">link</a>
> ```
> 
> If the page is at `/blog/2020/09/13/current-post`, then the href should point to `/blog/2020/09/13/another-post`, not `/another-post`. I.e. the resolver needs to be aware of the current URL, not just the site's URL.
> 
> Also, I would prefer not manually joining URLs in any case. Why not `elm.attribs.href = String(new URL(elm.attribs.href, currentPageURL))`?

Thanks for your comments. Sorry that I miss clicked the request review button.  I'll try to provide a better solution.

VinceCYLiao commented:
After reading the Nodejs's doc for the URL class, I found that the URL class itself can just handle this issue.
I have revised my code accordingly and built the website locally, and checked the hrefs in the generated feed files are resolved correctly.

@Josh-Cena  Please review at your convenience. Thank you!
 [feed.json of deploy preview](https://deploy-preview-9151--docusaurus-2.netlify.app/blog/feed.json)

VinceCYLiao commented:
Sorry that I forgot to run yarn test. Didn't noticed that so many tests will fail due to the changes. I'll look into how to fix the tests if you find my solution fine.

Josh-Cena commented:
I think it looks good! In fact the test changes look expected to me. I reckon you don't need to resolve paths that are just anchor links, only ones that actually point to another page. We would also need test cases (either adding a new test post file, or adding a link to the existing post file)

VinceCYLiao commented:
I encountered some problems while trying to add test cases.
A relative path without leading slash will make the build command throw.

I created two files:
website/_dogfooding/_blog tests/2023-07-19-a.mdx
website/_dogfooding/_blog tests/2023-07-19-b.mdx

Code in the 2023-07-19-a.mdx:
```md
 [link](b)
```

This work correctly in browser when running yarn start:website. the relative path "b" resolved into "/tests/blog/2023/07/19/b".
But while building the website, the handleBrokenLinks function will throw. And in the Error message, the relative path "b" is resolved into "/tests/b"
<img width="441" alt="image" src="https://github.com/facebook/docusaurus/assets/46563912/b756052d-d2a4-4459-b83d-44aeb6e452f9">

Is this considered a bug?

VinceCYLiao commented:
> I think it looks good! In fact the test changes look expected to me. I reckon you don't need to resolve paths that are just anchor links, only ones that actually point to another page. We would also need test cases (either adding a new test post file, or adding a link to the existing post file)

I have added a test case and updated the test plan. Please let me know if it's ok for you. Thanks!

Josh-Cena reviewed (changes requested):
Could you add the tests to https://github.com/facebook/docusaurus/tree/main/packages/docusaurus-plugin-content-blog/src/__tests__/__fixtures__/website instead? No one is going to look at the feed of the dogfooding blog.

VinceCYLiao commented:
I was thinking to parse the links from the feeds to check if they are correctly resolved, but found it's hard to so and maybe kind of meaningless since the links in feeds are from the object returned by the defaultCreateFeedItems function.
So my idea is to just write test case for the defaultCreateFeedItems function, checking if the full absolute path links are stay on touched while other links are correctly prefixed.
Don't know if the test makes sense for you, and if it is bad to export the defaultCreateFeedItems function just for unit test.

slorber commented on packages/docusaurus-plugin-content-blog/src/feed.ts:
```diff
@@ -106,6 +106,14 @@ async function defaultCreateFeedItems({
 
       const link = normalizeUrl([siteUrl, permalink]);
+
+      $(`div#${blogPostContainerID} a`).each((_, elm) => {
+        const {href} = elm.attribs;
+        if (href) {
+          elm.attribs.href = String(new URL(href, link));
+        }
+      });
```
LGTM 👍 

We will also need to convert image links to absolute, see 
https://github.com/facebook/docusaurus/issues/9136#issuecomment-1645593662

https://validator.w3.org/feed/docs/warning/ContainsRelRef.html

VinceCYLiao replied:
Done. Image links now are absolutized.

slorber commented on packages/docusaurus-plugin-content-blog/src/__tests__/feed.test.ts:
```diff
@@ -196,3 +228,95 @@ describe.each(['atom', 'rss', 'json'])('%s', (feedType) => {
   });
 });
+
+describe('Test defaultCreateFeedItems', () => {
```
That test looks super complex to me and I don't understand why.

Just call `createBlogFeedFiles` and take a snapshot: we'll review the snapshot and validate it contains what we expect

VinceCYLiao replied:
I think that I don't need to add new test case; instead I just need to update the snapshot. Am I correct ?

slorber commented on packages/docusaurus-plugin-content-blog/src/__tests__/feed.test.ts:
```diff
@@ -62,6 +68,32 @@ async function testGenerateFeeds(
   });
 }
 
+function isFullAbsolutePath(str: string) {
+  const domain = 'https://domain.com';
+  const {origin} = new URL(str, domain);
+  return origin !== domain;
+}
+
+async function generateLinksOfBlogPosts(outDir: string, blogPosts: BlogPost[]) {
+  const linksOfBlogPosts: {[postId: string]: string[]} = {};
+  const pathOfFile = path.join(outDir, 'blog');
+  const promises = blogPosts.map(async (post) => {
+    try {
+      const content = await readOutputHTMLFile(post.id, pathOfFile, true);
+      const $ = cheerioLoad(content);
+      const anchorElements = $(`div#${blogPostContainerID} a`);
+      if (anchorElements.length > 0) {
+        const href = anchorElements.map((_, elm) => elm.attribs.href).toArray();
+        linksOfBlogPosts[post.id] = href;
+      }
+    } catch {
+      // post is a draft
+    }
+  });
+  await Promise.all(promises);
+  return linksOfBlogPosts;
+}
```
I don't think we need that complexity inr our tests

VinceCYLiao replied:
The test case is removed

slorber reviewed (changes requested):
Thanks

The implementation looks good. 👍 

But the way it is tested looks surprisingly complex to me, and unit tests are not passing.

We also need to absolutize image URLs.

Let me know if you need help to figure out how to test that.

VinceCYLiao commented:
> Thanks
> 
> The implementation looks good. 👍
> 
> But the way it is tested looks surprisingly complex to me, and unit tests are not passing.
> 
> We also need to absolutize image URLs.
> 
> Let me know if you need help to figure out how to test that.

Thanks! I'll make the image URLs also absolutized.

Regarding the test, I'm thinking to create a new mock file which contains anchor elements with absolute/relative/anchor link, and image element with absolute/relative source URLs. And then in the test case just call createBlogFeedFiles and take a snapshot.

VinceCYLiao commented:
Tested in local and all unit tests are passed.

slorber commented on packages/docusaurus-plugin-content-blog/src/__tests__/__fixtures__/website/build-snap/blog/blog-with-links/index.html:
```diff
@@ -0,0 +1,31 @@
+<!doctype html>
```
How was this file generated?

VinceCYLiao replied:
I created the blog-with-links.mdx in the website/blog folder and ran yarn:build:website:blogOnly. If this is not the way how the files in __fixtures__ created, please let me know the create way to do it.

slorber commented on packages/docusaurus-plugin-content-blog/src/__tests__/__fixtures__/website/blog/blog-with-links.mdx:
```diff
@@ -0,0 +1,22 @@
+import dino from "../static/img/docusaurus.png";
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
```
ES imports are supposed to come after front matter

VinceCYLiao replied:
Order of imports are now correct. I also moved the front matter to the beginning of the file. Misplacing front matter seems to be the reason why tests failed.

slorber commented on packages/docusaurus-plugin-content-blog/src/feed.ts:
```diff
@@ -106,6 +106,33 @@ async function defaultCreateFeedItems({
+              elm.attribs.src = String(new URL(src, link));
+            }
+            if (srcset) {
+              elm.attribs.srcset = srcset
+                .split(',')
+                .map((s) => {
+                  const [imageURL, ...descriptors] = s.trim().split(/\s+/);
+                  const newImageURL = new URL(imageURL ?? '', link).href;
+                  return [newImageURL, ...descriptors].join(' ');
+                })
+                .join(', ');
```
Looks a bit unsafe/risky, maybe introduce a dedicated lib to manipulate srcset reliably instead? see https://www.npmjs.com/package/srcset

VinceCYLiao replied:
Thanks. I'll look into it and revise my code

VinceCYLiao replied:
Done and since latest version of srcset is pure ESM, so I have to use the previous version.

slorber reviewed (changes requested):
I don't undersand how it works anymore 😅 

@Josh-Cena do you remember what updates the `build-snap` folder exactly? Is this updated manually?

@VinceCYLiao how did this PR generate that new `src/__tests__/__fixtures__/website/build-snap/blog/blog-with-links/index.html` file?

The CI is failing and snapshots are not easy to review 😓

Josh-Cena commented:
> do you remember what updates the `build-snap` folder exactly? Is this updated manually?

I think I added this part of test but I don't remember how it works either. My guess is it's manual.

VinceCYLiao commented:
> I don't undersand how it works anymore 😅
> 
> @Josh-Cena do you remember what updates the `build-snap` folder exactly? Is this updated manually?
> 
> @VinceCYLiao how did this PR generate that new `src/__tests__/__fixtures__/website/build-snap/blog/blog-with-links/index.html` file?
> 
> The CI is failing and snapshots are not easy to review 😓 Surprisingly unit tests are passing locally, but not on GitHub action 🤷‍♂️

<img width="402" alt="image" src="https://github.com/facebook/docusaurus/assets/46563912/3eafbc40-242d-4a83-865b-26f2b5a27882">

Just ran the tests again and they are all passed. Sorry for the confusion and I'll look into why the tests are failing on github.

Josh-Cena reviewed (approved):
The implementation looks great to me! Just one stylistic suggestion.

slorber commented:
Hold on, I'm fixing the build-snap generation thing in another process before merging this PR

Josh-Cena commented:
Note that you would also want to rebase to get rid of the extra commits
